tests/test_cargo_metadata.rs

```diff
@@ -0,0 +1,89 @@
+use std::fs::File;
+use std::io::prelude::*;
+
+use support::{project, execs, basic_bin_manifest};
+use hamcrest::{assert_that, existing_file, is, equal_to};
+
+
+fn setup() {
+}
+
+test!(cargo_metadata_simple {
+    let p = project("foo")
+        .file("Cargo.toml", &basic_bin_manifest("foo"));
+
+    assert_that(p.cargo_process("metadata"), execs().with_stdout(format!(r#"
+[[packages]]
+dependencies = []
+manifest_path = "{}/Cargo.toml"
+name = "foo"
+version = "0.5.0"
+
+[[packages.targets]]
+kind = ["bin"]
+name = "foo"
+src_path = "src/foo.rs"
+
+[root]
+name = "foo"
+version = "0.5.0"
+
+"#, p.root().to_str().unwrap())));
+});
+
+test!(cargo_metadata_simple_json {
+    let p = project("foo")
+            .file("Cargo.toml", &basic_bin_manifest("foo"));
+
+    assert_that(p.cargo_process("metadata").arg("-f").arg("json"), execs().with_stdout(format!(
+    "{{\"root\":{{\"name\":\"foo\",\"version\":\"0.5.0\",\"features\":null}},\"packages\":\
+[{{\"name\":\"foo\",\"version\":\"0.5.0\",\"dependencies\":[],\"targets\":[{{\"kind\":[\"bin\"],\
+\"name\":\"foo\",\"src_path\":\"src/foo.rs\",\"metadata\":null}}],\"manifest_path\":\"{}/\
+Cargo.toml\"}}]}}\n", p.root().to_str().unwrap())));
```

Comment:
I think that it'll be helpful to make this a bit more readable (so it can change in that future). To that end I have a few suggestions:

* This shouldn't use `format!` as it requires many `{{` and `}}` characters, it's fine to just use a placeholder like `ROOT` and then `replace` the actually root into the string.
* This should use raw string syntax (`r"..."`) to avoid having to escape quotes
* This should have newlines to delimit output

In order to deal with ordering issues you may also want to just parse the output of the subcommand into JSON (and the expected JSON) and compare the two for equality.